### ISSUE TITLE
Use 2 spaces for unordered sublists, 3 spaces for ordered sublists

### ISSUE THREAD

**Issue by GandalfSaxe** [neilsustc/vscode-markdown#241]
#### What is the problem?
The feature of "Pressing Tab at the beginning of a list item will indent it" needs to be change to insert 2 spaces for unordered sublists, 3 spaces for ordered sublists. This is to be compatible CommonMark, and by extension the [markdownlint](https://github.com/DavidAnson/markdownlint) extension. It looks nicer in the source code because it's aligned.

See discussion at DavidAnson/markdownlint#137

#### How can I reproduce it?

```markdown
1. test
```
Press tab enter + tab when the cursor is at the end of the line above. It shouldn't simply insert X tabs, where X is the setting of the editor, it should insert exactly 3 spaces for reasons outlined above. Other dedicated markdown editors, such as [Typora](https://typora.io), already does this.

If you're worried that some users insist on non-CommonMark conforming indentations + doesn't care if the lists are nicely aligned and want the option to customize, then a new option could be introduced to override my suggested implementation, by following editor's space setting instead, as it does now.

**Comment by neilsustc** [comment]
Thanks for raising it. You can find more discussions at #155.

Personally, I would suggest "always 4 spaces scheme" rather than "2 and 3 spaces scheme" for unordered and ordered lists. One reason is that as a *code* editor, VSCode only recognizes one indentation setting for one file. "2 and 3 spaces" scheme confuses VSCode.

> ![image](https://user-images.githubusercontent.com/7588612/43054440-237a80dc-8e64-11e8-8ce1-09376dce66b2.png)

**Comment by GandalfSaxe** [comment]
So basically the choice is between:

### Solution 1 : Change to 4 spaces for all markdown sublist indentation:

**Pros:**
+ Indent guides are aligned.

**Cons:**
- Having to change `markdownlint` to allow 4 spaces.
- Not being CommonMark compliant.
- Not having ordered lists visually aligned


### Solution 2: Add option to `vscode-markdown` to use 2 spaces for UL, 3 spaces for UL

**Pros / Cons:**
Opposite of solution 1.

I'm willing to get used to anything. I don't have any strong preference regarding this. It's just that objectively, solution 2 seems to have less compromises, especially the CommonMark compliance one. Doesn't it count more to be in line with the greatest effort to standardize markdown, compared to a small detail around some visual guides?

If there is some other argument / perspective to this issue that I haven't seen, I'd be happy to know.

**Comment by neilsustc** [comment]
To my knowledge, *CommonMark don't specify the number of spaces we should use*. You can see it from the example in the very beginning of the spec

![image](https://user-images.githubusercontent.com/7588612/43066019-21637948-8e96-11e8-96e4-723e4ee2fcb3.png)

It uses mixed number of spaces to get a good visual appearance. (Although it do can use 3 here 😂)

So IMHO, **the rules for the number of spaces are only imposed by those lint tools, not the spec**.

Another case is having more than 9 list items. You cannot vertiaclly aligned them by still using 3 spaces then.
Don't ignore this case becuase someone does need 10+ items (#13)

---

I would modify your summary into these

### Solution 1: Always 4 spaces

**Pros**
- Indent guides are aligned
- Don't worry about the number of spaces, just keep 4 anywhere

**Cons**
- Having to change markdownlint to allow 4 spaces (I am not fully convinced by the current rules)
  > Another interesting example is [MD029](https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md029). It was `one` in the beginning, and I tried to be consistent with that although it is kind of unnatural (#54). (of course, we were back to `ordered` later)
  > Just now I found it was already changed to `one_or_ordered` ([commit](https://github.com/DavidAnson/markdownlint/commit/39d39db961bab0a0d67d49a4cf5928932ace817b)).
- ~~Not being CommonMark compliant~~ (I don't think so)
- Not having ordered lists visually aligned

### Solution 2: 2 and 3 spaces

**Pros / Cons**:
Opposite of solution 1.

But I would like to emphasis that "2 and 3 spaces scheme" won't make all of the people happy. Then it will become to "2, 3 and 4 spaces scheme"...

In conclusion, I would suggest to change the lint rule since it is not there in the CommonMark Spec.
And if you really cares about the visual appearance, I suggest adding one more space after ordered list marker

```
- one
- two
  - three

1.  one
2.  two
    3. three
```

**Comment by GandalfSaxe** [comment]
On further inspection, you're right about it being an issue with this particular `markdownlinting` rules, not the CommonMark spec. Apologies for the confusion. I agree with your pros/cons summary.

However, wouldn't it make everybody happy to:
1. Introduce option to use 2 spaces for UL, 3 for OL.
2. Keep default sublist spacing as it is now

**Comment by neilsustc** [comment]
Yes, it is completely doable.

For now the logic is to always indent with your configured `tabSize`
We can change it to looking for the line above and adopting the same indentation of it. I.e.
2 for <code>- item</code>,
3 for <code>1. item</code> and
4 for <code>1.&nbsp;&nbsp;item</code>

I think it is a good solution. If someone would like to help me, it could be great.

**Comment by GandalfSaxe** [comment]
Good!

I really wish I could help.

**Comment by linsui** [comment]
请问是否可以在格式化文件功能里加入根据 editor.tabSize 自动调整列表缩进的功能？

**Comment by linsui** [comment]
另一个请求是，目前当`editor.tabSize`为`true`时，在二级列表里退格会删除空格而非改变列表级别。

**Comment by neilsustc** [comment]
> 格式化文件时根据 editor.tabSize 自动调整列表缩进

这个用处不大，并且每个人喜欢不同的代码风格，这个 issue 本身就说明问题了

> 当 `editor.tabSize` 为 `true` 时，在二级列表里退格会删除空格而非改变列表级别

我不是很明白，`tabSize` 为 `true` 是什么设置😳，能再描述得详细一点吗，比如示例代码

（建议最好另外开一个 issue，不然会打扰别人）

**Comment by GandalfSaxe** [comment]
More reasons for this change of nested ordered list indented three spaces instead of two:

1. prettier.io also forces three spaces, i.e. 2 spaces or deleted down to 0 spaces, and 4 spaces and made into 3 spaces:

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBGAdAAgDYEsDOMAOlAExYDuuMAFiZg5hplHIXACb2PmbVwBbfCAA0ICAAcYuaMKSgAhgCclECgAVlCOSAUA3CLg6iQAIyUKwAazgwAyhMu4oAc2QwlAVzhjn+OEow6hYuAgrIAGYK2P5iAFb4AB4AQhbWtnYKAnAAMs5wkdGxII5K-krIIGFKVhxqUCYSSs4wAOpGtMgAHAAMYk0Q-q0WEpVNbAF6BWJKcACOnrizwQqh4UhRMT4g-gK47l7b+M4u2HAAip4Q8IVbYjAKpu0cnUik9xa4eK4AwhACYUqUGg0xAnn8ABVHnJNv4AL5woA

2. Also there are wildly more different behaviors for 2 spaces, and more agreement about 3 spaces across different markdown specs:

https://babelmark.github.io/?text=1.+list%0A2.+with%0A++1.+nested%0A++2.+items
vs
https://babelmark.github.io/?text=1.+list%0A2.+with%0A+++1.+nested%0A+++2.+items

**Comment by neilsustc** [comment]
For Prettier.io, it the same point "having a visually aligned look (con of 2-3 indentation)" as we summarised above.
And since it is a formatter (code beautifier), that is the job it needs to do 😂

For the second reason, actually, ordered list always requires at least 3 spaces, even in CommonMark Spec.
So it is not "different behaviors for 2 spaces" 😂, it is the same for most specs.
And that is why I recommend "always 4 spaces" and adding more spaces after marker to get them aligned.

**Comment by neilsustc** [comment]
You can now try out [the dev build](https://github.com/neilsustc/vscode-markdown#latest-development-build).
